Add networking support

raharper opened the pull request:
Implement the network connections, networks and nics configuration.

Network supports three types: user, bridge and mcast

'user'   - qemu slirp networking
'bridge' - host bridge and tun/tap interfaces (requires sudo privs)
'mcast'  - qemu socket networking using multicast routing between VMs
           using the same multicast network and ports (no sudo privs
           required)

raharper committed:
Add networking support

Implement the network connections, networks and nics configuration.

Network supports three types: user, bridge and mcast

'user'   - qemu slirp networking
'bridge' - host bridge and tun/tap interfaces (requires sudo privs)
'mcast'  - qemu socket networking using multicast routing between VMs
           using the same multicast network and ports (no sudo privs
           required)

hallyn commented on run.go:
```diff
@@ -106,6 +106,37 @@ func doRun(ctx *cli.Context) error {
+					return fmt.Errorf("Connection specified unknown network: %s", networkName)
+				}
+				for nidx := range vm.Nics {
+					found = false
```
This doesn't look right.  Did you mean the 'found = false' to happen outside of the for nidx := range vm.Nics loop?  Or am I misunderstanding the flow?

It looks like if vm.Nics is empty, you *should* return an error for the connection nic being undefined, but in fact 'found' will be true at line 134 and you'll proceed.

raharper replied:
I'll review.  I didn't test with a nic-less VMs, so that's likely a bug I'll fix.

raharper replied:
Thanks, found the issue.

hallyn commented on run.go:
```diff
@@ -114,11 +145,28 @@ func doRun(ctx *cli.Context) error {
+					nicNetwork = network
+					break
+				}
+			}
```
Can we assert here that nicNetwork was in fact found?

raharper replied:
Yes, I'll check it

hallyn commented on vmnics.go:
```diff
@@ -0,0 +1,185 @@
+// Allow specifying a prefix
+func randomNicIFName(prefix string) string {
+	prefixLength := len(prefix)
+	if prefixLength >= LinuxIfnameLength {
```
Shall we go a step further and require at least space for two random characters?  The pain of debugging subtle breakage when ifnames become likely to be the same is probably going to be worse than that of anyone wanting to use 14 character prefix.

raharper replied:
Sure

hallyn commented on vmnics.go:
```diff
@@ -0,0 +1,185 @@
+}
+
+// The VMNic lifecycle: first you create one with vm.AddNic().
+// The you call Setup on it to create the tap.  Then you call
```
*Then*

raharper replied:
+1

hallyn commented on vmnics.go:
```diff
@@ -0,0 +1,185 @@
+	return nil
+}
+
+var macOffset int = 0
```
Just a note that this works well so long as we continue to start clusters as a whole, but won't work if we start having separate programs start the nics.  (Unless we use an API to request a single daemon set up the hwaddrs)

raharper replied:
Yes, I think we should:

1) switch to starting clusters-as-a-whole via API
2) enable API to start/stop VMs within a cluster

hallyn replied:
Sounds good.

raharper committed:
Clean up handling of undefined networks or nics

raharper committed:
more fixes:

- randomNicName require at least two random chars for long prefix
- Assert nicNetwork was found after searching lists.
